### PR TITLE
Feature/Transformer ranking - Step 2-3

### PR DESCRIPTION
**Pull request by Jerry-Jinfeng-Guo**
Tentative 2-3, onto 4; SUBJECT to further CHANGES; not sure about `size_t` vs `int64_t` (cast)

N.B.: I got some clang-tidy errors in the `options.cpp` file:
error G1F71A4F3/G77134558/GA2D86B1A/G5B07D6B4/...: included header ***.hpp is not used directly

### CODE REVIEW

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp [hunk 1]
```diff
@@ -10,21 +10,30 @@
...
+#include <queue>
+#include <vector>
 
 namespace power_grid_model::optimizer {
 namespace tap_position_optimizer {
 
 namespace detail = power_grid_model::optimizer::detail;
 
-using TrafoGraphIdx = size_t;
+using TrafoGraphIdx = size_t; // Questionable: should we simply use Idx?
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:22]: i'm ok with it - not sure what boost uses internally, but it's probably a good idea to stick to their internal data type if they are sign-sensitive

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp [hunk 2]
```diff
@@ -10,21 +10,30 @@
...
 using EdgeWeight = int64_t;
+const int infty = INT_MAX;
+
+struct TrafoGraphVertex {
+    TrafoGraphIdx status{}; // status = 1 if the vertex is a source
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:27]: maybe `status` is the wrong name then? maybe make it `is_source`?

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp [hunk 3]
```diff
@@ -33,13 +42,63 @@ inline auto build_transformer_graph(State const& /*state*/) -> TransformerGraph
...
 
-inline auto get_edge_weights(TransformerGraph const& /*graph*/) -> std::vector<std::pair<Idx2D, EdgeWeight>> {
-    // TODO(mgovers): implement
-    return {};
+inline auto get_edge_weights(TransformerGraph const& graph) -> std::vector<std::pair<Idx2D, EdgeWeight>> {
+    // Step 2: Initialize the rank of all vertices (transformer nodes) as infinite (INT_MAX)
+    std::vector<EdgeWeight> rank(boost::num_vertices(graph), infty);
+    std::vector<Idx2D> sources(boost::num_vertices(graph));
+
+    // Step 3: Loop all the connected sources (status == 1)
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:50]: i think that this is quite obvious

```suggestion
    std::vector<EdgeWeight> rank(boost::num_vertices(graph), infty);
    std::vector<Idx2D> sources(boost::num_vertices(graph));

```

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp [hunk 4]
```diff
@@ -33,13 +42,63 @@ inline auto build_transformer_graph(State const& /*state*/) -> TransformerGraph
...
+    for (auto v : boost::make_iterator_range(boost::vertices(graph))) {
+        if (graph[v].status == 1) {
+            // a. Perform Dijkstra shortest path algorithm from the vertex with that source.
+            // This is to determine the shortest path of all vertices to this particular source.
+            std::priority_queue<std::pair<EdgeWeight, TrafoGraphIdx>, std::vector<std::pair<EdgeWeight, TrafoGraphIdx>>,
+                                std::greater<>>
+                pq;
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:57]: maybe create a `using` declaration for this to make it more readable (may be within this function as it's very specific)

```suggestion
            using TrafoGraphElement = std::pair<EdgeWeight, TrafoGraphIdx>;
            std::priority_queue<TrafoGraphElement, std::vector<TrafoGraphElement>, std::greater<>> pq;
```

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp [hunk 5]
```diff
@@ -33,13 +42,63 @@ inline auto build_transformer_graph(State const& /*state*/) -> TransformerGraph
...
+inline auto get_edge_weights(TransformerGraph const& graph) -> std::vector<std::pair<Idx2D, EdgeWeight>> {
+    // Step 2: Initialize the rank of all vertices (transformer nodes) as infinite (INT_MAX)
+    std::vector<EdgeWeight> rank(boost::num_vertices(graph), infty);
+    std::vector<Idx2D> sources(boost::num_vertices(graph));
+
+    // Step 3: Loop all the connected sources (status == 1)
+    for (auto v : boost::make_iterator_range(boost::vertices(graph))) {
+        if (graph[v].status == 1) {
+            // a. Perform Dijkstra shortest path algorithm from the vertex with that source.
+            // This is to determine the shortest path of all vertices to this particular source.
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:54]: please move this comment to the function docstring. it's very useful info, but not in this part of the function.

Alternatively, maybe move this sub-step to a separate function that does exactly do this step, again with this comment as the docstring

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp [hunk 6]
```diff
@@ -33,13 +42,63 @@ inline auto build_transformer_graph(State const& /*state*/) -> TransformerGraph
...
-    // TODO(mgovers): implement
-    return {};
+inline auto get_edge_weights(TransformerGraph const& graph) -> std::vector<std::pair<Idx2D, EdgeWeight>> {
+    // Step 2: Initialize the rank of all vertices (transformer nodes) as infinite (INT_MAX)
+    std::vector<EdgeWeight> rank(boost::num_vertices(graph), infty);
+    std::vector<Idx2D> sources(boost::num_vertices(graph));
+
+    // Step 3: Loop all the connected sources (status == 1)
+    for (auto v : boost::make_iterator_range(boost::vertices(graph))) {
+        if (graph[v].status == 1) {
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:52]: magic value. you can declare `constexpr Idx source_vertex = 1` and `constexpr Idx non_source_vertex = 0` earlier in the code (either outside where you initially declared the `status` or within this function)

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp [hunk 7]
```diff
@@ -33,13 +42,63 @@ inline auto build_transformer_graph(State const& /*state*/) -> TransformerGraph
...
-inline auto get_edge_weights(TransformerGraph const& /*graph*/) -> std::vector<std::pair<Idx2D, EdgeWeight>> {
-    // TODO(mgovers): implement
-    return {};
+inline auto get_edge_weights(TransformerGraph const& graph) -> std::vector<std::pair<Idx2D, EdgeWeight>> {
+    // Step 2: Initialize the rank of all vertices (transformer nodes) as infinite (INT_MAX)
+    std::vector<EdgeWeight> rank(boost::num_vertices(graph), infty);
+    std::vector<Idx2D> sources(boost::num_vertices(graph));
+
+    // Step 3: Loop all the connected sources (status == 1)
+    for (auto v : boost::make_iterator_range(boost::vertices(graph))) {
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:51]: wouldn't it be possible to do the entire graph in 1 go? right now, complexity is `V(E + V log V)` (cfr. wikipedia; performing a full Dijkstra for each vertex) but I feel like you can keep track of the current best path on the fly. At some point, for all vertices, their neighbours have `weight_neighbour >= weight_vertex` and you are done for all vertices

Comment (mgovers) [reply power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:51]: In `topology.hpp`, we use DFS (https://www.boost.org/doc/libs/1_84_0/libs/graph/doc/depth_first_search.html). My guess is that we can use BFS (https://cs.brown.edu/people/jwicks/boost/libs/graph/doc/breadth_first_search.html) to do this for us

**Comment by mgovers** [comment]
> N.B.: I got some clang-tidy errors in the options.cpp file:
> `error G1F71A4F3/G77134558/GA2D86B1A/G5B07D6B4/...: included header ***.hpp is not used directly`

never seen this error. can you show me the full message?

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp (hunk 5 above)
Comment (Jerry-Jinfeng-Guo) [reply power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:54]: Will be cleaned up once code is finalized. This is for now largely following a literate programming paradigm.

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp (hunk 7 above)
Comment (Jerry-Jinfeng-Guo) [reply power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:51]: Will revisit. Indeed room for optimization.

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp (hunk 3 above)
Comment (Jerry-Jinfeng-Guo) [reply power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:50]: Hmm, I do not quite follow this one

Comment (mgovers) [reply power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:50]: the comments don't add new information 😬 ; for instance: `initialize the rank of all vertices as infinite` is literally the next line. might as well leave it out in that case

**Comment by Jerry-Jinfeng-Guo** [comment]
> > N.B.: I got some clang-tidy errors in the options.cpp file:
> > `error G1F71A4F3/G77134558/GA2D86B1A/G5B07D6B4/...: included header ***.hpp is not used directly`
> 
> never seen this error. can you show me the full message?

This might be a local env thingy. I will share the log with you and if necessary let's make it a issue to investigate.

**Pull request merged**